src/coreclr/vm/stackwalk.cpp

```diff
@@ -415,10 +424,33 @@ UINT_PTR Thread::VirtualUnwindCallFrame(PREGDISPLAY pRD, EECodeInfo* pCodeInfo /
         PT_KNONVOLATILE_CONTEXT_POINTERS tempPtrs = pRD->pCurrentContextPointers;
         pRD->pCurrentContextPointers            = pRD->pCallerContextPointers;
         pRD->pCallerContextPointers             = tempPtrs;
+
+#ifdef TARGET_X86
+        pRD->PCTAddr = pRD->pCurrentContext->Esp - pCodeInfo->GetCodeManager()->GetStackParameterSize(pCodeInfo) - sizeof(DWORD);
+#endif
     }
     else
     {
+#ifdef TARGET_X86
+        GCInfoToken gcInfoToken = pCodeInfo->GetGCInfoToken();
+        hdrInfo hdrInfoBody;
+        DWORD hdrInfoSize = (DWORD)DecodeGCHdrInfo(gcInfoToken, pCodeInfo->GetRelOffset(), &hdrInfoBody);
+
+        ::UnwindStackFrameX86(pRD,
+                            PTR_CBYTE(pCodeInfo->GetSavedMethodCode()),
+                            pCodeInfo->GetRelOffset(),
+                            &hdrInfoBody,
+                            dac_cast<PTR_CBYTE>(gcInfoToken.Info) + hdrInfoSize,
+                            PTR_CBYTE(pCodeInfo->GetJitManager()->GetFuncletStartAddress(pCodeInfo)),
+                            pCodeInfo->IsFunclet(),
+                            true);
+
+        pRD->pCurrentContext->ContextFlags |= CONTEXT_UNWOUND_TO_CALL;
+        pRD->pCurrentContext->Esp = pRD->SP;
+        pRD->pCurrentContext->Eip = pRD->ControlPC;
+#else
         VirtualUnwindCallFrame(pRD->pCurrentContext, pRD->pCurrentContextPointers, pCodeInfo);
```

Comment:
What is the benefit of having the unwinding code expanded here instead of using the `VirtualUnwindCallFrame` that seems to have the same code?

Reply:
Not needing to redecode GC info to get PCTAddr value.

Reply:
I can see the GC info decoding in both methods, so I am not sure I understand how it helps.

Reply:
The second one doesn't give back PCTAddr (part of REGDISPLAY but not part of CONTEXT) value. In order to recompute it we would need to either pass it back somehow or recompute it, which requires GC info decoding.

Reply:
The `pCodeInfo->GetCodeManager()->GetStackParameterSize` call in the case above does GC info decoding. We could do that in both cases but then you will get one GC info decoding inside the `VirtualUnwindCallFrame(T_CONTEXT*, T_KNONVOLATILE_CONTEXT_POINTERS*, EECodeInfo*)` call and another one from computing the `PCTAddr` using  `GetStackParameterSize`.

Reply:
I don't see the first one setting the PCTAddr, maybe it is missing then?

Reply:
`::UnwindStackFrameX86` sets it internally by calling `SetRegdisplayPCTAddr`. (As a side node, I tried also a version that included `pEip` in `T_KNONVOLATILE_CONTEXT_POINTERS` but the change is unnecessarily big due to frame handling cgenx86.cpp)